### DISCUSSION TITLE
Additional sources for Hong Kong

### DISCUSSION THREAD

**Discussion by chunshek** [breezy-weather/breezy-weather#1417] [Ideas]
### Describe your idea

Following the [conversation](https://github.com/breezy-weather/breezy-weather/pull/1405#issuecomment-2466821254) at the end of the BMKG PR:

> If Hong Kong can access other weather sources with no issues (firewall), then we can suggest other sources in the location preset. Wdyt?

Hong Kong is **not** behind the GFW of China (yet), so we can indeed suggest other sources (air quality, pollen, precipitation nowcasting) in `LocationPreset.kt`.

### Potential air quality source
Rather than falling back to `openmeteo`, we could use official air pollutant monitoring data from the Environmental Protection Department:
<https://www.aqhi.gov.hk/epd/ddata/html/out/24pc_Eng.xml>

There are a few reasons why I hadn't included in my HKO PR though:
1. It is issued from the EPD, not from HKO. So it should be included as a separate source (like MF and Recosanté).
2. It is in XML. I would have to spend time to study how the XML parser is used in other parts of Breezy Weather. I wanted to get HKO out of the gate first.
3. The data refers to the 18 stations by name only, so the station coordinates need to be hardcoded – and I know your general preference is to avoid doing that. (There is an [official source](https://portal.csdi.gov.hk/geoportal/?datasetId=epd_rcd_1629267205214_40635) for the station coordinates too, but it's a ZIP archive of an FGDB / GeoJSON file, which was last updated in January 2023. Downloading and uncompressing it everytime Breezy Weather pulls air quality data seems to be a waste of an effort to implement unzip, using up redundant bandwidth for every update, etc.)

If you are fine with hardcoding the station coordinates for the EPDHK source though, I can aim for the next release (after 5.3.0).

### Acknowledgements

- [X] I have searched the [existing issues](https://github.com/breezy-weather/breezy-weather/issues) and [existing discussions](https://github.com/breezy-weather/breezy-weather/discussions/categories/ideas) and this is a new ticket, **NOT** a duplicate or related to another open OR closed issue or discussion.
- [X] The idea I’m suggesting is not in the [list of features that will not be implemented](https://github.com/breezy-weather/breezy-weather/blob/main/README.md#features-that-will-not-be-implemented).
- [X] I have updated the app to **[version 5.2.8 or later](https://github.com/breezy-weather/breezy-weather/releases/latest)**. Even if I’m getting updates from F-Droid default repo, I don’t need to uninstall/reinstall to update.
- [X] I wrote my request in English, and any screenshot I attached are also in English (language can be changed from Settings > Appearance). I can write in my native language under the English text, if necessary.
- [X] I understand that this app is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY, so I may not get a positive reply to my request, if any at all.

**Comment by papjul** [comment]
I thought we had station coordinates in Breezy Weather? In the end, they were all removed?

I found the grid coordinates, but I wonder if we can't simply replace them with a check on `countryCode`?

**Reply by chunshek** [reply]
Okay. Let me break it down. (I know I've contributed several sources by now and it's not easy to keep tabs.)
- CWA has about 30 station coordinates hardcoded for temperature normals in Taiwan.
- HKO does not have observation station coordinates hardcoded.
- HKO has two grids (don't ask why):
  - The 16×15 "forecast grid" is evenly spaced (0.1° × 0.1°). The exact grid rectangle can be determined with simple arithmetic.
  - The 18×18 "current grid" is unevenly spaced. We need to identify which grid rectangle a location belongs to, so as to retrieve the observation from the correct station. There is a 100KB+ [static GeoJSON file](https://my.weather.gov.hk/hiking/geojson/grid.geojson) that defines the grid. We could download it every time we need to get location parameters. Or we could simply hardcode the 38 boundary lines (19 horizontal and 19 vertical).
- EPD does not report to HKO. They are separate. EPD stations are not placed with any respect of either HKO grids.

**Reply by papjul** [reply]
Do you have a representation of these grids? I mean, if they just cover Hong Kong, it might just be easier to check for the country code.

**Reply by chunshek** [reply]
Sure! Here's HKO's "current grid" overlaid on top of the district map of Hong Kong (sorry didn't load up our beautiful coastline).
![image](https://github.com/user-attachments/assets/88e0863b-1a5c-4423-9831-ec5fa7f170d0)

Each of the grid boxes is served by a pre-defined set of nearby weather stations for delivering current observations. For instance, the current observations of G0906 can be retrieved from <https://my.weather.gov.hk/PDADATA/locspc/data/gridData/0906_en.xml> (which is a JSON file despite the `.xml` extension):

```
{
    "RegionalWeather": {
        "Grid": "0906",
        "ObsTime": "202411112340",
        "DisplayTime": "2024\/11\/11 23:40",
        "Temp": {
            "Value": "24.1",
            "MinTemp": "24",
            "MaxTemp": "28",
            "Temperature_AroundtoOdd": "24",
            "ModelMaxTemperature": "28",
            "ModelMinTemperature": "24",
            "Station": "HKO",
            "DefaultStation": "HKO",
            "Valid": true
        },
        "RH": {
            "Value": "82",
            "Station": "HKO",
            "Valid": true
        },
        "Wind": {
            "WindSpeed": "10",
            "Gust": "19",
            "WindDirection": "East",
            "WindDirectionCode": "E",
            "WindDirectionAbbr": "E",
            "Station": "SF",
            "Valid": true
        },
        "Pressure": {
            "Value": "1014.3",
            "Station": "HKO",
            "Valid": true
        }
    },
    "Rainfall": {
        "ObsTime": "202411112330",
        "GridMapping": "RF_ANAL[8][5][0]",
        "Value": "0"
    },
    "WeatherPhoto": {
        "photo": "HK2"
    }
}
```

Therefore, for all locations within the G0906 box, the current measurements for temperature, humidity, and pressure are taken from the `HKO` station, while winds are from the `SF` station. Locations in another box would show current measurements from a different combination of stations.

> I mean, if they just cover Hong Kong, it might just be easier to check for the country code.

I am not sure what checking for the country code would achieve. 373 of these 384 boxes fall within Hong Kong's territorial boundary. Different grid boxes would get different observations, even if they are all within HK country code.

**Reply by chunshek** [reply]
The current grid ID is also used for reverse geolocation within Hong Kong. For example, if a location falls within the G0906 box, we match it against the polygons stored in <https://my.weather.gov.hk/PDADATA/locspc/data/gridData/geojson/L0906.geojson> to identify its name.

Because of how densely the current grid is spaced near the city center, each of these GeoJSON files ends up very small and quick to process.

![image](https://github.com/user-attachments/assets/ab5828d6-2459-4032-a969-f949745b11f6)
